src/avatar.cpp

```diff
@@ -459,183 +457,180 @@ bool avatar::read( item &it, const bool continuous )
-                }
-            }
-            if( !nonlearners.empty() ) {
-                add_header( _( "Not participating:" ) );
-                for( const auto &elem : nonlearners ) {
-                    menu.addentry( -1, false, -1, get_text( nonlearners, elem ) );
-                }
-            }
+    int learner_id = -1;
+
+    //only show the menu if there's useful information or multiple options
+    if( skill || !nonlearners.empty() || !fun_learners.empty() ) {
+        uilist menu;
+
+        // Some helpers to reduce repetition:
+        auto length = []( const std::pair<npc *, std::string> &elem ) {
+            return utf8_width( elem.first->disp_name() ) + utf8_width( elem.second );
+        };
+
+        auto max_length = [&length]( const std::map<npc *, std::string> &m ) {
+            auto max_ele = std::max_element( m.begin(),
+                                             m.end(), [&length]( const std::pair<npc *, std::string> &left,
+            const std::pair<npc *, std::string> &right ) {
+                return length( left ) < length( right );
+            } );
+            return max_ele == m.end() ? 0 : length( *max_ele );
+        };
+
+        auto get_text =
+        [&]( const std::map<npc *, std::string> &m, const std::pair<npc *, std::string> &elem ) {
+            const int lvl = elem.first->get_skill_level( skill );
+            const std::string lvl_text = skill ? string_format( _( " | current level: %d" ), lvl ) : "";
+            const std::string name_text = elem.first->disp_name() + elem.second;
+            return string_format( "%s%s", left_justify( name_text, max_length( m ) ), lvl_text );
+        };
+
+        auto add_header = [&menu]( const std::string & str ) {
+            menu.addentry( -1, false, -1, "" );
+            uilist_entry header( -1, false, -1, str, c_yellow, c_yellow );
+            header.force_color = true;
+            menu.entries.push_back( header );
+        };
+
+        menu.title = !skill ? string_format( _( "Reading %s" ), book->type_name() ) :
+                     //~ %1$s: book name, %2$s: skill name, %3$d and %4$d: skill levels
+                     string_format( _( "Reading %1$s (can train %2$s from %3$d to %4$d)" ), book->type_name(),
+                                    skill_name, type->req, type->level );
+
 
-            menu.query( true );
-            if( menu.ret == UILIST_CANCEL ) {
-                add_msg( m_info, _( "Never mind." ) );
-                return false;
+        menu.addentry( 0, true, '0', _( "Read once" ) );
+
+        const int lvl = get_skill_level( skill );
+        menu.addentry( 2 + getID().get_value(), lvl < type->level, '1',
+                       string_format( _( "Read until you gain a level | current level: %d" ), lvl ) );
+
+
+        if( !learners.empty() ) {
+            add_header( _( "Read until this NPC gains a level:" ) );
+            for( const std::pair<npc *, std::string> &elem : learners ) {
```

Comment:
```suggestion
            for( const std::pair<npc *const, std::string> &elem : learners ) {
```
Not sure if the first `const` needs removing.